engineapi/proxy.go

```diff
@@ -0,0 +1,312 @@
+package engineapi
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	imclient "github.com/longhorn/longhorn-instance-manager/pkg/client"
+	imtypes "github.com/longhorn/longhorn-instance-manager/pkg/types"
+	imutil "github.com/longhorn/longhorn-instance-manager/pkg/util"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/types"
+)
+
+const (
+	ProxyPollInterval = 5 * time.Second
+)
+
+type EngineClientProxy interface {
+	EngineClient
+	BackupTargetBinaryClient
+
+	IsGRPC() bool
+	Start(*longhorn.InstanceManager, logrus.FieldLogger, *datastore.DataStore) error
+	Stop(string) error
+	Ping() error
+}
+
+type EngineClientProxyHandler struct {
+	Clients map[string]EngineClientProxy
+
+	logger logrus.FieldLogger
+
+	ds *datastore.DataStore
+
+	lock *sync.RWMutex
+}
+
+func NewEngineClientProxyHandler(logger logrus.FieldLogger, ds *datastore.DataStore) *EngineClientProxyHandler {
+	return &EngineClientProxyHandler{
+		Clients: make(map[string]EngineClientProxy),
+		logger:  logger,
+		ds:      ds,
+		lock:    &sync.RWMutex{},
+	}
+}
+
+func getLoggerForProxy(logger logrus.FieldLogger, image string) *logrus.Entry {
+	return logger.WithFields(
+		logrus.Fields{
+			"instanceManagerImage": image,
+		},
+	)
+}
+
+func (h *EngineClientProxyHandler) GetClient(im *longhorn.InstanceManager, fallBack interface{}) (c EngineClientProxy, err error) {
+	image := ""
+	if im == nil {
+		image, err = h.ds.GetSettingValueExisted(types.SettingNameDefaultInstanceManagerImage)
```

Comment:
Then you need to get the IM CR for this default image. Otherwise, the below `h.NetClient` must fail and the caller has no way to get the client.

Reply:
Not expecting im to be nil. Removed.